Datastream API

LucasSaragosa opened the pull request:
Initial barebone design. Added support for file/memory streams, URLs and temp files.

Supported platforms:
- [x] Windows
- [X] MacOS
- [x] Linux (TBD)

LucasSaragosa committed:
datastream API

initial barebone design. platform specific updates NEEDED for win/linux. file/memory streams, manager, urls. temp files support.

LucasSaragosa committed:
Update Platform.cpp

fix return

LucasSaragosa committed:
Update .gitignore

LucasSaragosa committed:
Update .gitignore

LucasSaragosa committed:
remcmake

LucasSaragosa committed:
fixes

fix crc32 and other fixes

iMrShadow committed:
Add Windows datastream API functionality

iMrShadow committed:
Fix std::replace not found

iMrShadow committed:
Format code using our clang rules

LucasSaragosa committed:
Update Platform.cpp

update windows filesize

LucasSaragosa committed:
Update Platform.cpp

fix

LucasSaragosa committed:
Update Platform.cpp

remove commented code

LucasSaragosa committed:
scripting and context

Scripting manager comments. Lua manager comments. Lua manager delegates lua funcionality. cmake build folder changes to Bin/. global library context type created used as a global struct to manage modding.

iMrShadow commented on .github/workflows/cmake-multi-platform.yml:
Why was this removed?

asilz committed:
Add linux implementations

iMrShadow commented on CMakeLists.txt:
```diff
@@ -16,8 +16,12 @@ if(PROJECT_SOURCE_DIR STREQUAL PROJECT_BINARY_DIR) # Prevent in-source builds
 
 set(TELLTALE_EDITOR_ROOT_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
-set(CMAKE_INSTALL_PREFIX "${TELLTALE_EDITOR_ROOT_DIR}/bin")
+set(CMAKE_INSTALL_PREFIX "${TELLTALE_EDITOR_ROOT_DIR}/BinInstall")
```
I was ok with ./bin, because that's auto-generated and not part of the project.

iMrShadow commented on TelltaleEditor/Source/ToolLibrary/Include/Resource/DataStream.hpp:
```diff
@@ -0,0 +1,209 @@
+
+    // Constructs with the symbol scheme. The resource URL that points to that symbol. If GetScheme() is still symbol, it was not found.
+    // You can still call open and another find attempt is made, so Open() may succeed if it now exists.
+    // nIf it is not this, then it was found and you can open it.
```
Small typo

LucasSaragosa committed:
lua scripts

initial meta game script. data streams buffer, sub stream.

LucasSaragosa committed:
Merge branch 'datastream-integration' of https://github.com/Telltale-Modding-Group/Telltale-Editor into datastream-integration

iMrShadow commented on TelltaleEditor/Source/Editor/Source/Main.cpp:
```diff
@@ -7,13 +7,7 @@ int main()
-            TTE_LOG("Could not load meta.lua");
-        else
-            ScriptManager::RunText(Context.GetLibraryLVM(), src);
+        Context.Switch({"TX100","PC",""});
```
TX100 is too vague, but this can be changed at a later date.

LucasSaragosa replied:
i think its good for most games. like for mc story mode it should be MCSM100 and MC200 for s2, because thats the format telltale use for seasons and games (their apple package name in the .app is com.telltalegames.mc100 etc

iMrShadow replied:
I was specifically talking about TX100 to be renamed, for e.g. TEXAS100, otherwise it would sound like a model name from the Terminator.

LucasSaragosa replied:
lmao. ids are internal though and should be as concise as possible

iMrShadow replied:
I know they are for internal usage, but they still confuse me. I have something like [this](https://gitflic.ru/project/pashok6798/ttg-tools/blob?file=master%2FMainMenu.cs&branch=master#line-num-69) on my mind, but a little bit better.

LucasSaragosa committed:
Update Platform.cpp

update linux seek

LucasSaragosa commented:
@iMrShadow please add windows FilePos and FileSeek

iMrShadow committed:
Reverse deletion of cmake-multi-platform yml script

iMrShadow committed:
Update comment

iMrShadow committed:
Format code

iMrShadow committed:
Add Windows FilePos and FileSeek functionality

iMrShadow committed:
Enable GetPosition and SetPosition support, reorganize code, fix a small typo

LucasSaragosa committed:
game init

initialise meta scripts

iMrShadow commented on TelltaleEditor/Source/ToolLibrary/Source/Meta/Meta.cpp:
```diff
@@ -1,14 +1,50 @@
+
+#include <map>
+
+static ToolContext* GlobalContext = nullptr;
```
Prefer using the singleton pattern.

iMrShadow commented on TelltaleEditor/Source/ToolLibrary/Source/Meta/Meta.cpp:
```diff
@@ -18,6 +54,14 @@ namespace Meta {
+            
+            String v = ScriptManager::PopString(man);
+            
+            reg.MetaVersion = v == "MSV6" ? MSV6 : v == "MSV5" ? MSV5 : v == "MSV4" ? MSV4 : v == "MCOM" ? MCOM : v == "MTRE" ? MTRE : v == "MBIN" ? MBIN : v == "MBES" ? MBES : (StreamVersion)-1;
```
Prefer using a switch case. It can also be extracted to a different helper function.

LucasSaragosa replied:
switch only works for POD and strings arent D:

iMrShadow replied:
The day when pattern matching comes to C++, I will be a happy man. In the meantime, this can be extracted to a "GetMetaVersion" function with multiple if/elseifs.

iMrShadow commented on TelltaleEditor/Source/Editor/Source/Main.cpp:
```diff
@@ -5,10 +5,11 @@ int main()
     {
         
-        ToolContext Context{};
+        ToolContext& Context = *CreateToolContext();
```
```cpp
ToolContext* Context = CreateToolContext();
Context->DoSmth();
```

This can be converted to a singleton.

iMrShadow commented on Dev/Games.lua:
```diff
@@ -3,4 +3,5 @@
 local texasHoldem = {}
 texasHoldem.Name = "Telltale Texax Hold'em"
 texasHoldem.ID = "TX100" -- Texas S1 (100)
+texasHoldem.MetaVersion = "MBIN" -- Version of meta streams which files created will be
```
Meta versions need to be researched thoroughly, since some games have mixed ones. Strong Bad's Cool Game for Attractive People had 2 episodes with MTRE, while the others were using MBIN. Retail versions may have an updated engine. 

Meta versions could be determined at runtime?

LucasSaragosa replied:
by this is mean the default version for new files. any version, in any game can be loaded. perhaps i should rename to DefaultMetaVersion

iMrShadow replied:
Maybe a comment would suffice.

LucasSaragosa replied:
yea

iMrShadow committed:
Update Main.cpp

iMrShadow committed:
Improve codebase

- Format code using our clang-format rules
- Change some documentation
- Convert some macros to constexprs
